File: RunCoach/RunCoach/Models/Enums.swift
import Foundation

enum RunningAbility: String, CaseIterable, Codable {
    case beginner, intermediate, advanced

    var label: String {
        switch self {
        case .beginner: return "Beginner"
        case .intermediate: return "Intermediate"
        case .advanced: return "Advanced"
        }
    }

    var description: String {
        switch self {
        case .beginner: return "Running less than 6 months or under 15 mi/week"
        case .intermediate: return "1–3 years of consistent running, 15–35 mi/week"
        case .advanced: return "3+ years, racing regularly, 35+ mi/week"
        }
    }
}

enum RunningGoal: String, CaseIterable, Codable {
    case getFit, loseWeight, fiveK, tenK, halfMarathon, marathon, fasterMile

    var label: String {
        switch self {
        case .getFit: return "Get Fit"
        case .loseWeight: return "Lose Weight"
        case .fiveK: return "Run a 5K"
        case .tenK: return "Run a 10K"
        case .halfMarathon: return "Half Marathon"
        case .marathon: return "Marathon"
        case .fasterMile: return "Faster Mile"
        }
    }

    var emoji: String {
        switch self {
        case .getFit: return "💪"
        case .loseWeight: return "⚖️"
        case .fiveK: return "🏁"
        case .tenK: return "🎽"
        case .halfMarathon: return "🏅"
        case .marathon: return "🏆"
        case .fasterMile: return "⚡️"
        }
    }

    var peakMileageRange: ClosedRange<Double> {
        switch self {
        case .getFit, .loseWeight: return 20...30
        case .fiveK: return 20...30
        case .tenK: return 25...35
        case .halfMarathon: return 35...45
        case .marathon: return 45...55
        case .fasterMile: return 20...30
        }
    }

    var hasRaceDistance: Bool {
        switch self {
        case .fiveK, .tenK, .halfMarathon, .marathon: return true
        default: return false
        }
    }
}

enum TrainingStyle: String, CaseIterable, Codable {
    case conservative, balanced, aggressive

    var label: String {
        switch self {
        case .conservative: return "Conservative"
        case .balanced: return "Balanced"
        case .aggressive: return "Aggressive"
        }
    }

    var description: String {
        switch self {
        case .conservative: return "Safety first. Slower progression, more rest. Best for injury-prone runners."
        case .balanced: return "Standard approach. Steady progress with adequate recovery."
        case .aggressive: return "Push your limits. Faster progression for experienced, healthy runners."
        }
    }

    var weeklyIncreasePercent: Double {
        switch self {
        case .conservative: return 0.05
        case .balanced: return 0.08
        case .aggressive: return 0.10
        }
    }
}

enum WorkoutType: String, CaseIterable, Codable {
    case easy, long, tempo, intervals, strides, rest, strength, mobility

    var label: String {
        switch self {
        case .easy: return "Easy Run"
        case .long: return "Long Run"
        case .tempo: return "Tempo Run"
        case .intervals: return "Intervals"
        case .strides: return "Easy + Strides"
        case .rest: return "Rest"
        case .strength: return "Strength"
        case .mobility: return "Mobility"
        }
    }

    var emoji: String {
        switch self {
        case .easy: return "🏃"
        case .long: return "🛣️"
        case .tempo: return "⏱️"
        case .intervals: return "⚡️"
        case .strides: return "💨"
        case .rest: return "😴"
        case .strength: return "🏋️"
        case .mobility: return "🧘"
        }
    }

    var isHard: Bool {
        self == .tempo || self == .intervals
    }

    var isRun: Bool {
        self == .easy || self == .long || self == .tempo || self == .intervals || self == .strides
    }

    var color: String {
        switch self {
        case .easy: return "green"
        case .long: return "blue"
        case .tempo: return "orange"
        case .intervals: return "red"
        case .strides: return "teal"
        case .rest: return "gray"
        case .strength: return "purple"
        case .mobility: return "mint"
        }
    }
}

enum InjuryType: String, CaseIterable, Codable {
    case none, knee, itBand, shinSplints, plantarFasciitis, hip, back, hamstring, other

    var label: String {
        switch self {
        case .none: return "None"
        case .knee: return "Knee"
        case .itBand: return "IT Band"
        case .shinSplints: return "Shin Splints"
        case .plantarFasciitis: return "Plantar Fasciitis"
        case .hip: return "Hip"
        case .back: return "Back"
        case .hamstring: return "Hamstring"
        case .other: return "Other"
        }
    }
}

enum TrainingPhase: String, Codable {
    case base, build, peak, taper, deload

    var label: String {
        switch self {
        case .base: return "Base"
        case .build: return "Build"
        case .peak: return "Peak"
        case .taper: return "Taper"
        case .deload: return "Recovery"
        }
    }

    var description: String {
        switch self {
        case .base: return "Building your aerobic engine with easy miles"
        case .build: return "Adding quality workouts and increasing mileage"
        case .peak: return "Your highest volume and intensity weeks"
        case .taper: return "Reducing load so you arrive fresh on race day"
        case .deload: return "Easy recovery week to absorb training stress"
        }
    }
}

File: RunCoachRN/src/types/enums.ts
export enum RunningAbility {
  Beginner = 'beginner',
  Intermediate = 'intermediate',
  Advanced = 'advanced',
}

export enum RunningGoal {
  GetFit = 'getFit',
  FiveK = 'fiveK',
  TenK = 'tenK',
  HalfMarathon = 'halfMarathon',
  Marathon = 'marathon',
  FasterMile = 'fasterMile',
  Ultra50 = 'ultra50',
  Ultra100 = 'ultra100',
}

export enum TrainingStyle {
  Conservative = 'conservative',
  Balanced = 'balanced',
  Aggressive = 'aggressive',
}

export enum WorkoutType {
  Easy = 'easy',
  Long = 'long',
  Tempo = 'tempo',
  Intervals = 'intervals',
  Strides = 'strides',
  Rest = 'rest',
  Strength = 'strength',
  Mobility = 'mobility',
  Walk = 'walk',
}

export enum InjuryType {
  None = 'none',
  Knee = 'knee',
  ITBand = 'itBand',
  ShinSplints = 'shinSplints',
  PlantarFasciitis = 'plantarFasciitis',
  Hip = 'hip',
  Back = 'back',
  Hamstring = 'hamstring',
  Other = 'other',
}

export enum TrainingPhase {
  Base = 'base',
  Build = 'build',
  Peak = 'peak',
  Taper = 'taper',
  Deload = 'deload',
}

export enum InjuryRisk {
  Low = 'low',
  Medium = 'medium',
  High = 'high',
}

export enum ExperienceLevel {
  Beginner = 'beginner',
  Intermediate = 'intermediate',
  Advanced = 'advanced',
}

export enum CompletionStatus {
  Completed = 'completed',
  Partial = 'partial',
  Skipped = 'skipped',
}

export enum PainLevel {
  None = 'none',
  Mild = 'mild',
  Moderate = 'moderate',
  Sharp = 'sharp',
}

export enum EnergyLevel {
  Low = 'low',
  Normal = 'normal',
  High = 'high',
}

export enum SleepQuality {
  Poor = 'poor',
  Ok = 'ok',
  Good = 'good',
}

export enum SignalLevel {
  Ok = 'ok',
  Caution = 'caution',
  Concern = 'concern',
  Danger = 'danger',
}

// MARK: - Label / display helpers

export const ABILITY_LABELS: Record<RunningAbility, string> = {
  [RunningAbility.Beginner]: 'Beginner',
  [RunningAbility.Intermediate]: 'Intermediate',
  [RunningAbility.Advanced]: 'Advanced',
};

export const ABILITY_DESCRIPTIONS: Record<RunningAbility, string> = {
  [RunningAbility.Beginner]: 'Running less than 6 months or under 15 mi/week',
  [RunningAbility.Intermediate]: '1–3 years of consistent running, 15–35 mi/week',
  [RunningAbility.Advanced]: '3+ years, racing regularly, 35+ mi/week',
};

export const GOAL_LABELS: Record<RunningGoal, string> = {
  [RunningGoal.GetFit]: 'Get Fit',
  [RunningGoal.FiveK]: 'Run a 5K',
  [RunningGoal.TenK]: 'Run a 10K',
  [RunningGoal.HalfMarathon]: 'Half Marathon',
  [RunningGoal.Marathon]: 'Marathon',
  [RunningGoal.FasterMile]: 'Faster Mile',
  [RunningGoal.Ultra50]: '50-Mile Ultra',
  [RunningGoal.Ultra100]: '100-Mile Ultra',
};

export const GOAL_EMOJIS: Record<RunningGoal, string> = {
  [RunningGoal.GetFit]: '💪',
  [RunningGoal.FiveK]: '🏁',
  [RunningGoal.TenK]: '🎽',
  [RunningGoal.HalfMarathon]: '🏅',
  [RunningGoal.Marathon]: '🏆',
  [RunningGoal.FasterMile]: '⚡️',
  [RunningGoal.Ultra50]: '🌄',
  [RunningGoal.Ultra100]: '🦅',
};

export const WORKOUT_LABELS: Record<WorkoutType, string> = {
  [WorkoutType.Easy]: 'Easy Run',
  [WorkoutType.Long]: 'Long Run',
  [WorkoutType.Tempo]: 'Tempo Run',
  [WorkoutType.Intervals]: 'Intervals',
  [WorkoutType.Strides]: 'Easy + Strides',
  [WorkoutType.Rest]: 'Rest',
  [WorkoutType.Strength]: 'Strength',
  [WorkoutType.Mobility]: 'Mobility',
  [WorkoutType.Walk]: 'Walk',
};

export const WORKOUT_EMOJIS: Record<WorkoutType, string> = {
  [WorkoutType.Easy]: '🏃',
  [WorkoutType.Long]: '🛣️',
  [WorkoutType.Tempo]: '⏱️',
  [WorkoutType.Intervals]: '⚡️',
  [WorkoutType.Strides]: '💨',
  [WorkoutType.Rest]: '😴',
  [WorkoutType.Strength]: '🏋️',
  [WorkoutType.Mobility]: '🧘',
  [WorkoutType.Walk]: '🚶',
};

export const PHASE_LABELS: Record<TrainingPhase, string> = {
  [TrainingPhase.Base]: 'Base',
  [TrainingPhase.Build]: 'Build',
  [TrainingPhase.Peak]: 'Peak',
  [TrainingPhase.Taper]: 'Taper',
  [TrainingPhase.Deload]: 'Recovery',
};

export const INJURY_LABELS: Record<InjuryType, string> = {
  [InjuryType.None]: 'None',
  [InjuryType.Knee]: 'Knee',
  [InjuryType.ITBand]: 'IT Band',
  [InjuryType.ShinSplints]: 'Shin Splints',
  [InjuryType.PlantarFasciitis]: 'Plantar Fasciitis',
  [InjuryType.Hip]: 'Hip',
  [InjuryType.Back]: 'Back',
  [InjuryType.Hamstring]: 'Hamstring',
  [InjuryType.Other]: 'Other',
};

export const TRAINING_STYLE_LABELS: Record<TrainingStyle, string> = {
  [TrainingStyle.Conservative]: 'Conservative',
  [TrainingStyle.Balanced]: 'Balanced',
  [TrainingStyle.Aggressive]: 'Aggressive',
};

export const TRAINING_STYLE_DESCRIPTIONS: Record<TrainingStyle, string> = {
  [TrainingStyle.Conservative]: 'Safety first. Slower progression, more rest. Best for injury-prone runners.',
  [TrainingStyle.Balanced]: 'Standard approach. Steady progress with adequate recovery.',
  [TrainingStyle.Aggressive]: 'Push your limits. Faster progression for experienced, healthy runners.',
};

// MARK: - Type helpers

export function isHardWorkout(type: WorkoutType): boolean {
  return type === WorkoutType.Tempo || type === WorkoutType.Intervals;
}

export function isRunWorkout(type: WorkoutType): boolean {
  return [WorkoutType.Easy, WorkoutType.Long, WorkoutType.Tempo,
          WorkoutType.Intervals, WorkoutType.Strides, WorkoutType.Walk].includes(type);
}

export function painRank(level: PainLevel): number {
  const ranks: Record<PainLevel, number> = {
    [PainLevel.None]: 0,
    [PainLevel.Mild]: 1,
    [PainLevel.Moderate]: 2,
    [PainLevel.Sharp]: 3,
  };
  return ranks[level];
}

export function completionWeight(status: CompletionStatus): number {
  const weights: Record<CompletionStatus, number> = {
    [CompletionStatus.Completed]: 1.0,
    [CompletionStatus.Partial]: 0.5,
    [CompletionStatus.Skipped]: 0.0,
  };
  return weights[status];
}

export function goalPeakMileageRange(goal: RunningGoal): [number, number] {
  const ranges: Record<RunningGoal, [number, number]> = {
    [RunningGoal.GetFit]: [20, 30],
    [RunningGoal.FiveK]: [20, 30],
    [RunningGoal.TenK]: [25, 35],
    [RunningGoal.HalfMarathon]: [35, 45],
    [RunningGoal.Marathon]: [45, 55],
    [RunningGoal.FasterMile]: [20, 30],
    [RunningGoal.Ultra50]:  [55, 70],
    [RunningGoal.Ultra100]: [65, 80],
  };
  return ranges[goal];
}

export function goalMaxLongRunMiles(goal: RunningGoal): number {
  const max: Record<RunningGoal, number> = {
    [RunningGoal.GetFit]: 10,
    [RunningGoal.FiveK]: 8,
    [RunningGoal.TenK]: 10,
    [RunningGoal.HalfMarathon]: 14,
    [RunningGoal.Marathon]: 22,
    [RunningGoal.FasterMile]: 8,
    [RunningGoal.Ultra50]:  32,
    [RunningGoal.Ultra100]: 40,
  };
  return max[goal];
}

export function goalNeedsTaper(goal: RunningGoal): boolean {
  return [RunningGoal.HalfMarathon, RunningGoal.Marathon,
          RunningGoal.Ultra50, RunningGoal.Ultra100].includes(goal);
}

export function isUltraGoal(goal: RunningGoal): boolean {
  return goal === RunningGoal.Ultra50 || goal === RunningGoal.Ultra100;
}

export function injuryRiskMaxIncrease(risk: InjuryRisk, base: number): number {
  const factors: Record<InjuryRisk, number> = {
    [InjuryRisk.Low]: 1.10,
    [InjuryRisk.Medium]: 1.07,
    [InjuryRisk.High]: 1.05,
  };
  return base * factors[risk];
}

export function experienceUsesPaceZones(level: ExperienceLevel): boolean {
  return level !== ExperienceLevel.Beginner;
}
